### PR TITLE
TNO-2756 Fix Pie Chart

### PR DESCRIPTION
**Pull request by Fosol**
Pie charts were having issues with the latest configuration options.  We can now configure them to handle different scenarios.  There is still an undesirable result when including more than one dataset, however we probably shouldn't bother with this type.

We can now control the line colour in charts, and apply an array of colours to each group value.

## Summary

- Add DB migration 1.0.180
- Publish tno-core:0.1.88
- Update Editor app
- Update Subscriber app
- Update API
- Update Charts API

### CODE REVIEW

**Comment by kkwangsir** [comment]
I also have a migration underway, I will wait for you guys to merge first.

File: api/node/src/utils/data.ts [hunk 1]
```diff
@@ -35,15 +35,17 @@ export const evalScripts = (
       const dataset = data.datasets[i];
       try {
         const bkColor = dataset.backgroundColor;
-        if (bkColor && typeof bkColor === 'string' && bkColor[0] === '(')
+        const isArray = Array.isArray(bkColor);
```
Comment (Fosol) [inline api/node/src/utils/data.ts:38]: Colour can not be an array

File: app/editor/src/features/admin/charts/components/ChartTemplatePreviewConfig.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import { Col, mergeChartSettings, Row, ToggleButton } from 'tno-core';
+
+import { useChartTemplateContext } from '../ChartTemplateContext';
+import * as styled from './styled';
+
+/**
+ * The page used to view and edit a chart template.
+ * @returns Component.
+ */
+export const ChartTemplatePreviewConfig = () => {
```
Comment (Fosol) [inline app/editor/src/features/admin/charts/components/ChartTemplatePreviewConfig.tsx:19]: Provides a way to view the current config and edit it

File: app/editor/src/features/admin/charts/templates/chartTemplate.cshtml [hunk 3]
```diff
@@ -11,12 +12,21 @@
...
-    $"\"spanGaps\": true, " +
-    $"\"borderWidth\": 2, " +
-    $"\"minBarLength\": {SectionSettings.MinBarLength?.ToString() ?? "null"}, " +
-    $"\"borderColor\": \"{ReportExtensions.GetColor(SectionSettings.DatasetColors, datasetIndex++, SectionSettings.Dataset, ds.Key, SectionSettings.DatasetValue)}\" }}"))]
+  "datasets": [@String.Join(", ", datasets.Select((ds) => {
+    var index = datasetIndex++;
+    var indexValue = SectionSettings.ApplyColorToValue ? -1 : index;
+    var data = ReportExtensions.ExtractDatasetValues(SectionSettings.DatasetValue, SectionSettings.GroupBy, ds, labels);
+    var backgroundColors = ReportExtensions.GetColors(SectionSettings.DatasetColors, indexValue, SectionSettings.Dataset, ds.Key, SectionSettings.DatasetValue).Select((v) => "\"" + v + "\"");
+    var borderColors = ReportExtensions.GetColors(SectionSettings.DatasetBorderColors, indexValue, SectionSettings.Dataset, ds.Key, SectionSettings.DatasetValue).Select((v) => "\"" + v + "\"");
```
Comment (Fosol) [inline app/editor/src/features/admin/charts/templates/chartTemplate.cshtml:20]: Separate border colours

File: app/editor/src/features/admin/charts/templates/chartTemplate.cshtml [hunk 4]
```diff
@@ -11,12 +12,21 @@
...
-    $"\"backgroundColor\": \"{ReportExtensions.GetColor(SectionSettings.DatasetColors, datasetIndex, SectionSettings.Dataset, ds.Key, SectionSettings.DatasetValue)}\", " +
-    $"\"spanGaps\": true, " +
-    $"\"borderWidth\": 2, " +
-    $"\"minBarLength\": {SectionSettings.MinBarLength?.ToString() ?? "null"}, " +
-    $"\"borderColor\": \"{ReportExtensions.GetColor(SectionSettings.DatasetColors, datasetIndex++, SectionSettings.Dataset, ds.Key, SectionSettings.DatasetValue)}\" }}"))]
+  "datasets": [@String.Join(", ", datasets.Select((ds) => {
+    var index = datasetIndex++;
+    var indexValue = SectionSettings.ApplyColorToValue ? -1 : index;
+    var data = ReportExtensions.ExtractDatasetValues(SectionSettings.DatasetValue, SectionSettings.GroupBy, ds, labels);
+    var backgroundColors = ReportExtensions.GetColors(SectionSettings.DatasetColors, indexValue, SectionSettings.Dataset, ds.Key, SectionSettings.DatasetValue).Select((v) => "\"" + v + "\"");
```
Comment (Fosol) [inline app/editor/src/features/admin/charts/templates/chartTemplate.cshtml:19]: We now support an array of colours.

File: libs/net/models/Settings/ReportSectionSettingsModel.cs [hunk 5]
```diff
@@ -19,8 +19,10 @@ public class ReportSectionSettingsModel
     public string GroupBy { get; set; } = "";
     public string SortBy { get; set; } = "";
     public string SortDirection { get; set; } = "";
-    public string[] DatasetColors { get; set; } = Array.Empty<string>();
-    public string[] DataLabelColors { get; set; } = Array.Empty<string>();
+    // public bool ApplyColorToValue { get; set; }
```
Comment (Fosol) [inline libs/net/models/Settings/ReportSectionSettingsModel.cs:22]: I've removed, as I don't think they are used anymore.

File: libs/net/template/Models/Charts/ChartDatasetModel.cs [hunk 6]
```diff
@@ -24,13 +24,13 @@ public class ChartDatasetModel
     /// get/set -
     /// </summary>
     [JsonPropertyName("backgroundColor")]
-    public string? BackgroundColor { get; set; }
+    public string[]? BackgroundColor { get; set; }
```
Comment (Fosol) [inline libs/net/template/Models/Charts/ChartDatasetModel.cs:27]: Technically both string and array are supported by Chart.js, but we can't do this in .NET.  Javascript is garbage.

**Pull request merged**